### PR TITLE
DSND-423-unit tests to ensure retryable/non-retryable exceptions thrown

### PR DESCRIPTION
**Pull request by amobarak**
Following exceptions must be raised (Acceptance criterias) - Comment added to each unit test for all criterias:
image

Avro Deserialization already tested in deserializer/serializer

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java [hunk 1]
```diff
@@ -67,6 +70,58 @@ void When_ValidChsDeltaMessage_Expect_ValidInsolvencyDeltaMapping() throws IOExc
...
+        final ApiResponse<Void> response = new ApiResponse<>(HttpStatus.BAD_REQUEST.value(), null, null);
+
+        when(transformer.transform(expectedInsolvency)).thenReturn(internalCompanyInsolvencyMock());
+        when(apiClientService.putInsolvency(eq("context_id"),eq("02588581"), eq(internalCompanyInsolvencyMock()))).thenReturn(response);
+        Assertions.assertThrows(NonRetryableErrorException.class, () -> deltaProcessor.processDelta(mockChsDeltaMessage));
+        verify(apiClientService).putInsolvency("context_id", "02588581", internalCompanyInsolvencyMock());
+        verify(transformer).transform(expectedInsolvency);
+    }
+
+    @Test
```
Comment (amobarak) [inline src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java:88]: image
Throws retryable error

File: src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java [hunk 2]
```diff
@@ -67,6 +70,58 @@ void When_ValidChsDeltaMessage_Expect_ValidInsolvencyDeltaMapping() throws IOExc
...
+        final ApiResponse<Void> response = new ApiResponse<>(HttpStatus.UNAUTHORIZED.value(), null, null);
+
+        when(transformer.transform(expectedInsolvency)).thenReturn(internalCompanyInsolvencyMock());
+        when(apiClientService.putInsolvency(eq("context_id"),eq("02588581"), eq(internalCompanyInsolvencyMock()))).thenReturn(response);
+        Assertions.assertThrows(RetryableErrorException.class, () -> deltaProcessor.processDelta(mockChsDeltaMessage));
+        verify(apiClientService).putInsolvency("context_id", "02588581", internalCompanyInsolvencyMock());
+        verify(transformer).transform(expectedInsolvency);
+    }
+
+    @Test
```
Comment (amobarak) [inline src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java:103]: image
Any 5xx Response, returns retryable exception

File: src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java [hunk 3]
```diff
@@ -67,6 +70,58 @@ void When_ValidChsDeltaMessage_Expect_ValidInsolvencyDeltaMapping() throws IOExc
...
+        final ApiResponse<Void> response = new ApiResponse<>(HttpStatus.INTERNAL_SERVER_ERROR.value(), null, null);
+
+        when(transformer.transform(expectedInsolvency)).thenReturn(internalCompanyInsolvencyMock());
+        when(apiClientService.putInsolvency(eq("context_id"),eq("02588581"), eq(internalCompanyInsolvencyMock()))).thenReturn(response);
+        Assertions.assertThrows(RetryableErrorException.class, () -> deltaProcessor.processDelta(mockChsDeltaMessage));
+        verify(apiClientService).putInsolvency("context_id", "02588581", internalCompanyInsolvencyMock());
+        verify(transformer).transform(expectedInsolvency);
+    }
+
+    @Test
```
Comment (amobarak) [inline src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java:118]: image
Delta transformation into API

File: src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java [hunk 4]
```diff
@@ -47,60 +48,60 @@ public InsolvencyDeltaProcessor(ApiClientService apiClientService,
...
-            final String partition =
-                    Objects.requireNonNull(
-                            headers.get(KafkaHeaders.RECEIVED_PARTITION_ID)
-                    ).toString();
-            final String offset =
-                    Objects.requireNonNull(headers.get(KafkaHeaders.OFFSET)).toString();
-
-            ObjectMapper mapper = new ObjectMapper();
-            InsolvencyDelta insolvencyDelta = mapper.readValue(payload.getData(),
+            insolvencyDelta = mapper.readValue(payload.getData(),
```
Comment (zahidm3) [inline src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java:67]: Shouldn't de-serialisation failure here be a Non-Retryable exception as per story?

File: src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java [hunk 5]
```diff
@@ -67,6 +70,58 @@ void When_ValidChsDeltaMessage_Expect_ValidInsolvencyDeltaMapping() throws IOExc
...
+    @DisplayName("When can't transform into insolvency delta API, throws retryable error")
```
Comment (amobarak) [inline src/test/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessorTest.java:119]: image
JSON deserializer throws non retryable exception

File: src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java [hunk 6]
```diff
@@ -47,59 +48,66 @@ public InsolvencyDeltaProcessor(ApiClientService apiClientService,
...
+
+    }
+
+    private InsolvencyDelta mapToInsolvencyDelta(ChsDelta payload)
+            throws NonRetryableErrorException {
+        ObjectMapper mapper = new ObjectMapper();
+        try {
+            return mapper.readValue(payload.getData(),
+                    InsolvencyDelta.class);
+        } catch (JsonProcessingException exception) {
```
Comment (gseregni-ch) [inline src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java:108]: The readValue method cold throws:
com.fasterxml.jackson.core.JsonProcessingException
com.fasterxml.jackson.databind.JsonMappingException

Please catch also the JsonMappingException

And for both we should throw NonRetryableException.

File: src/main/java/uk/gov/companieshouse/insolvency/delta/consumer/InsolvencyDeltaConsumer.java [hunk 7]
```diff
@@ -56,7 +56,12 @@ public void receiveMainMessages(Message<ChsDelta> message,
                                     @Header(KafkaHeaders.OFFSET) String offset) {
         logger.trace(String.format("A new message from %s topic with payload:%s "
                 + "and headers:%s ", topic, message.getPayload(), message.getHeaders()));
-        deltaProcessor.processDelta(message);
+        try {
+            deltaProcessor.processDelta(message);
+        } catch (Exception exception) {
+            logger.error(exception);
```
Comment (gseregni-ch) [inline src/main/java/uk/gov/companieshouse/insolvency/delta/consumer/InsolvencyDeltaConsumer.java:62]: Please log a meaningful message in front of the exception. Something like "Exception while processing the {topic-name} - message in receiveMainMessages method"

File: src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java [hunk 8]
```diff
@@ -47,59 +48,66 @@ public InsolvencyDeltaProcessor(ApiClientService apiClientService,
...
-            InsolvencyDelta insolvencyDelta = mapper.readValue(payload.getData(),
-                    InsolvencyDelta.class);
-            logger.trace(String.format("DSND-362: InsolvencyDelta extracted "
+        MessageHeaders headers = chsDelta.getHeaders();
+        final ChsDelta payload = chsDelta.getPayload();
+        final String logContext = payload.getContextId();
+        final Map<String, Object> logMap = new HashMap<>();
+        final String receivedTopic =
+                Objects.requireNonNull(headers.get(KafkaHeaders.RECEIVED_TOPIC)).toString();
+        final String partition =
```
Comment (gseregni-ch) [inline src/main/java/uk/gov/companieshouse/insolvency/delta/processor/InsolvencyDeltaProcessor.java:57]: Topic, partition and offset are part of the header of the receiveMainMessages method.
We can probably pass them to this method instead than retrieve them from the KafkaHeaders object.

**Review by zahidm3 — APPROVED** [review]

**Review by gseregni-ch — APPROVED** [review]

**Pull request merged**